### PR TITLE
server: refactor listNetworks api database retrievals

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

This PR attempts to optimize database queries during listNetworks API call. It should fix pagination-related issues as well while querying desired networks from the database. Only in the following cases, it may query all networks:
- API param `canusefordeploy` is passed as true
- Any service specified with the `supportedservices` API parameter

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
good idea

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
In my testing, listing networks is working fine with the changes and smoke tests didn't report any error. Marking it ready for review

**Review by rohityadavcloud — APPROVED** [review]
LGTM, didn't test it but needs testing on:
- networks by roles: admin, domain admin, normal user
- list networks by projects, accounts, domains
- list networks in env with variety of networks across roles; isolated networks, VPC tier/network, shared network, L2 network and user-shared network

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@sureshanaparti if possible, I would prefer to have this in 4.19 (either 4.19.1 or 4.19.2)
@blueorangutan test

**Comment by vishesh92** [comment]
I created an isolated & a L2 network. I got the below error when I tried to add a networkfilter=shared.
```
(localcloud) __ > list networks filter=id,name networkfilter=shared                                                                                                                                                                           
__ Error: (HTTP 530, error code 4250) DB Exception on: com.mysql.cj.jdbc.ClientPreparedStatement: SELECT networks.id, networks.mode, networks.broadcast_domain_type, networks.traffic_type, networks.name, networks.display_text, networks.bro
adcast_uri, networks.gateway, networks.cidr, networks.network_cidr, networks.network_offering_id, networks.vpc_id, networks.physical_network_id, networks.data_center_id, networks.related, networks.guru_name, networks.state, networks.redun
dant, networks.domain_id, networks.account_id, networks.set_fields, networks.guru_data, networks.dns1, networks.dns2, networks.ip6Dns1, networks.ip6Dns2, networks.network_domain, networks.removed, networks.created, networks.reservation_id
, networks.uuid, networks.guest_type, networks.acl_type, networks.restart_required, networks.specify_ip_ranges, networks.ip6_gateway, networks.ip6_cidr, networks.display_network, networks.network_acl_id, networks.streched_l2, networks.ext
ernal_id, networks.public_mtu, networks.private_mtu FROM networks  INNER JOIN network_offerings networkOfferingSearch ON networks.network_offering_id=networkOfferingSearch.id  INNER JOIN domain domainSearch ON networks.domain_id=domainSea
rch.id  INNER JOIN data_center zoneSearch ON networks.data_center_id=zoneSearch.id  INNER JOIN account accountSearch ON networks.account_id=accountSearch.id WHERE networks.removed IS NULL  AND networks.display_network = 1  AND  ()  AND ne
tworks.removed IS NULL  AND  (networkOfferingSearch.system_only = 0 ) ORDER BY networks.id DESC  LIMIT 0, 500 
```

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 1]
```diff
@@ -2324,111 +2332,121 @@ public Pair<List<? extends Network>, Integer> searchForNetworks(ListNetworksCmd
...
+        SearchCriteria<NetworkVO> additionalSearchCriteria = _networksDao.createSearchCriteria();
+
         if (isSystem == null || !isSystem) {
             if (!permittedAccounts.isEmpty()) {
                 if (Arrays.asList(Network.NetworkFilter.Account, Network.NetworkFilter.AccountDomain, Network.NetworkFilter.All).contains(networkFilter)) {
                     //get account level networks
-                    networksToReturn.addAll(listAccountSpecificNetworks(buildNetworkSearchCriteria(sb, keyword, id, isSystem, zoneId, guestIpType, trafficType, physicalNetworkId, networkOfferingId,
-                            aclType, skipProjectNetworks, restartRequired, specifyIpRanges, vpcId, tags, display, vlanId, associatedNetworkId), searchFilter, permittedAccounts));
+                    additionalSearchCriteria.addOr("id", SearchCriteria.Op.SC,
+                            getAccountSpecificNetworksSearchCriteria(sb, permittedAccounts, skipProjectNetworks));
```
Comment (vishesh92) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:2345]: If `getAccountSpecificNetworksSearchCriteria` returns null, the request fails with a bad sql query.

Comment (shwstppr) [reply server/src/main/java/com/cloud/network/NetworkServiceImpl.java:2345]: Thanks @vishesh92 will check

Comment (shwstppr) [reply server/src/main/java/com/cloud/network/NetworkServiceImpl.java:2345]: @vishesh92 it was due to empty searchcriteria at line 2392. Should be fixed now.

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
Should a user role be able to search for networks services? Like for example list all networks that support specific service (like dns, DHCP etc)?

**Review by borisstoyanov — APPROVED** [review]
LGTM, checked manually for admin/user visibility of shared networks and for networks deployed within projects

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
Merging based on manual tests & reviews.

**Pull request merged**